### PR TITLE
stat: define ROC cutoffs instead of using number of cutoffs

### LINKED ISSUE #316: stat: Change parity convention in ROC
I was going through the ROC AUC examples and found that the results were not very intuitive. Indeed [the basic example](https://github.com/gonum/gonum/blob/master/stat/roc_example_test.go#L41) gives a ROC AUC of 0.75 but if I run it with sklearn I get:

```python
>>> from sklearn import metrics
>>> y = [0.1, 0.35, 0.4, 0.8]
>>> classes = [1, 0, 1, 0]
>>> metrics.roc_auc_score(classes, y)
0.25
```

Moreover, according to gonum a perfect match between two slices results in a ROC AUC of 0, whereas intuitively it should be 1. Initially I made [a pull request](https://github.com/gonum/gonum/pull/315) because I thought it was a typo in the examples but it turns out the problem is a bit more complicated.

### LINKED ISSUE #708: stat: ROC signature
I came up with what I think is an improvement to the ROC function, 
in which the cutoffs are defined explicitly instead of the number of cutoffs n.
This allows for unequally-spaced cutoffs, allows for all cutoffs to be used
more naturally, and simplifies the code itself.

Is this a good idea -- should I put a PR in for it, @sbinet @kortschak ?

Also, I was thinking maybe ROC should output an error as well, instead of all
the panic statements? Or is it appropriate that it panic?

What I did is here, in case you wanted to look at the specifics of what I am 
suggesting:
https://github.com/Armadilloa16/stat/tree/roc

### LINKED ISSUE #960: stat: add thresholds return value to ROC function
This matches sklearn's `roc_curve` function.

### PR DESCRIPTION
**Pull request by kortschak**
Please take a look.

Closes #708.
Fixes #316.
Closes #960.

### CODE REVIEW

**Comment by kortschak** [comment]
The fix in #408 was wrong. The test cases should have been changed rather than the behaviour.

**Comment by kortschak** [comment]
Pulling out the class inversion changes. They only superficially matches sklearn's behaviour.

**Comment by kortschak** [comment]
Now the behaviour of `ROC` pretty much exactly matches sklearn's `metrics.roc_curve`.

The diff to do comparison with sklearn is this.
```diff
diff --git a/stat/roc_test.go b/stat/roc_test.go
index 36c658b..65af285 100644
--- a/stat/roc_test.go
+++ b/stat/roc_test.go
@@ -5,7 +5,11 @@
 package stat
 
 import (
+	"fmt"
 	"math"
+	"os"
+	"os/exec"
+	"strings"
 	"testing"
 
 	"gonum.org/v1/gonum/floats"
@@ -220,5 +224,62 @@ func TestROC(t *testing.T) {
 		if !floats.Same(gotThresh, test.wantThresh) {
 			t.Errorf("%d: unexpected thresholds got:%#v want:%v", i, gotThresh, test.wantThresh)
 		}
+
+		if len(test.cutoffs) != 0 {
+			continue
+		}
+		sklearn(test.y, test.c, test.w)
+		fmt.Println(gotTPR, gotFPR, gotThresh)
+		fmt.Println()
+	}
+}
+
+func sklearn(y []float64, c []bool, w []float64) {
+	tmpl := `from sklearn import metrics
+y = %s
+c = %s
+w = %s
+print(y, c, w)
+fpr, tpr, thresh = metrics.roc_curve(y_score=y, y_true=c, sample_weight=w)
+print(tpr, fpr, thresh)`
+	if len(w) == 0 {
+		w = make([]float64, len(y))
+		for i := range w {
+			w[i] = 1
+		}
+	}
+	snippet := fmt.Sprintf(tmpl, pyFloat(y), pyBool(c), pyFloat(w))
+	cmd := exec.Command("python", "-c", snippet)
+	cmd.Stdout = os.Stdout
+	cmd.Run()
+}
+
+func pyFloat(f []float64) string {
+	var buf strings.Builder
+	buf.WriteByte('[')
+	for i, v := range f {
+		if i != 0 {
+			buf.WriteString(", ")
+		}
+		fmt.Fprintf(&buf, "%f", v)
+	}
+	buf.WriteByte(']')
+	return buf.String()
+}
+
+func pyBool(b []bool) string {
+	var buf strings.Builder
+	buf.WriteByte('[')
+	for i, v := range b {
+		if i != 0 {
+			buf.WriteString(", ")
+		}
+		if v {
+			buf.WriteByte('1')
+		} else {
+			buf.WriteByte('0')
+		}
 	}
+	buf.WriteByte(']')
+	return buf.String()
 }
```

**Comment by kortschak** [comment]
Output from the diffed code above:
```
([0.0, 3.0, 5.0, 6.0, 7.5, 8.0], [0, 1, 0, 1, 1, 1], [1.0, 1.0, 1.0, 1.0, 1.0, 1.0])
(array([0.25, 0.75, 0.75, 1.  , 1.  ]), array([0. , 0. , 0.5, 0.5, 1. ]), array([8., 6., 5., 3., 0.]))
[0 0.25 0.5 0.75 0.75 1 1] [0 0 0 0 0.5 0.5 1] [+Inf 8 7.5 6 5 3 0]

([0.0, 3.0, 5.0, 6.0, 7.5, 8.0], [0, 1, 0, 1, 1, 1], [4.0, 1.0, 6.0, 3.0, 2.0, 2.0])
(array([0.25 , 0.5  , 0.875, 0.875, 1.   , 1.   ]), array([0. , 0. , 0. , 0.6, 0.6, 1. ]), array([8. , 7.5, 6. , 5. , 3. , 0. ]))
[0 0.25 0.5 0.875 0.875 1 1] [0 0 0 0 0.6 0.6 1] [+Inf 8 7.5 6 5 3 0]

([0.0, 3.0, 6.0, 6.0, 6.0, 8.0], [0, 1, 0, 1, 1, 1], [1.0, 1.0, 1.0, 1.0, 1.0, 1.0])
(array([0.25, 0.75, 1.  , 1.  ]), array([0. , 0.5, 0.5, 1. ]), array([8., 6., 3., 0.]))
[0 0.25 0.75 1 1] [0 0 0.5 0.5 1] [+Inf 8 6 3 0]

([0.0, 3.0, 6.0, 6.0, 6.0, 8.0], [0, 1, 0, 1, 1, 1], [4.0, 1.0, 6.0, 3.0, 2.0, 2.0])
(array([0.25 , 0.875, 1.   , 1.   ]), array([0. , 0.6, 0.6, 1. ]), array([8., 6., 3., 0.]))
[0 0.25 0.875 1 1] [0 0 0.6 0.6 1] [+Inf 8 6 3 0]

([0.1, 0.35, 0.4, 0.8], [1, 0, 1, 0], [1.0, 1.0, 1.0, 1.0])
(array([0. , 0. , 0.5, 0.5, 1. ]), array([0. , 0.5, 0.5, 1. , 1. ]), array([1.8 , 0.8 , 0.4 , 0.35, 0.1 ]))
[0 0 0.5 0.5 1] [0 0.5 0.5 1 1] [+Inf 0.8 0.4 0.35 0.1]

([0.1, 0.35, 0.4, 0.8], [0, 0, 1, 1], [1.0, 1.0, 1.0, 1.0])
(array([0.5, 1. , 1. ]), array([0., 0., 1.]), array([0.8, 0.4, 0.1]))
[0 0.5 1 1 1] [0 0 0 0.5 1] [+Inf 0.8 0.4 0.35 0.1]

([1.0, 2.0], [0, 0], [1.0, 1.0])
(array([nan, nan, nan]), array([0. , 0.5, 1. ]), array([3., 2., 1.]))
[NaN NaN NaN] [0 0.5 1] [+Inf 2 1]

([1.0], [0], [1.0])
(array([nan, nan]), array([0., 1.]), array([2., 1.]))
[NaN NaN] [0 1] [+Inf 1]

([1.0], [1], [1.0])
(array([1.]), array([nan]), array([1.]))
[0 1] [NaN NaN] [+Inf 1]

([], [], [])
[] [] []
```

File: stat/roc_example_test.go [hunk 1]
```diff
@@ -6,51 +6,106 @@ package stat_test
...
+	tpr, fpr, _ := stat.ROC(nil, y, classes, nil)
 	fmt.Printf("true  positive rate: %v\n", tpr)
 	fmt.Printf("false positive rate: %v\n", fpr)
 
 	// Output:
-	// true  positive rate: [0 0.5 0.5 1 1 1 1]
-	// false positive rate: [0 0 0.25 0.25 0.5 0.75 1]
+	// true  positive rate: [0 0.25 0.5 0.75 0.75 1 1]
+	// false positive rate: [0 0 0 0 0.5 0.5 1]
+}
```
Comment (sbinet) [inline stat/roc_example_test.go]: perhaps also add an example where the thresholds are displayed?

perhaps one that matches this one from `sklearn.metrics.roc_curve`?
https://scikit-learn.org/stable/modules/generated/sklearn.metrics.roc_curve.html
```python
>>> import numpy as np
>>> from sklearn import metrics
>>> y = np.array([1, 1, 2, 2])
>>> scores = np.array([0.1, 0.4, 0.35, 0.8])
>>> fpr, tpr, thresholds = metrics.roc_curve(y, scores, pos_label=2)
>>> fpr
array([0. , 0. , 0.5, 0.5, 1. ])
>>> tpr
array([0. , 0.5, 0.5, 1. , 1. ])
>>> thresholds
array([1.8 , 0.8 , 0.4 , 0.35, 0.1 ])
```

Comment (kortschak) [reply stat/roc_example_test.go]: Done

**Pull request merged**